crypto/cipher_extra/e_aes_cbc_hmac_sha1.c

```diff
@@ -0,0 +1,362 @@
+
+  key->payload_length = NO_PAYLOAD_LENGTH;
+
+  if (len % AES_BLOCK_SIZE) {
+    OPENSSL_PUT_ERROR(CIPHER, CIPHER_R_DATA_NOT_MULTIPLE_OF_BLOCK_LENGTH);
+    return 0;
+  }
+
+  if (EVP_CIPHER_CTX_encrypting(ctx)) {
+    // NOTE: Difference between openssl and aws-lc:
+    // In encrypt case, |plen| is set in the call |EVP_CIPHER_CTX_ctrl| with
+    // |EVP_CTRL_AEAD_TLS1_AAD| operation.
+    // When |plen == NO_PAYLOAD_LENGTH|, it means the call did not happen.
+    // In this case, aws-lc returns error(0) but openssl supports that with
+    // below explanation.
+    // https://mta.openssl.org/pipermail/openssl-users/2019-November/011458.html
+    // -- These stitched ciphers are specifically targeted at use by libssl
+    //    and are designed for use in SSL/TLS only.
+    if (plen == NO_PAYLOAD_LENGTH) {
+      // |EVP_CIPHER_CTX_ctrl| with |EVP_CTRL_AEAD_TLS1_AAD| operation is not
+      // performed.
+      OPENSSL_PUT_ERROR(CIPHER, CIPHER_R_CTRL_OPERATION_NOT_PERFORMED);
+      return 0;
+    }
+    if (len !=
+        ((plen + SHA_DIGEST_LENGTH + AES_BLOCK_SIZE) & -AES_BLOCK_SIZE)) {
+      // The input should have space for eiv + plaintext + digest + padding.
+      OPENSSL_PUT_ERROR(CIPHER, CIPHER_R_UNSUPPORTED_INPUT_SIZE);
+      return 0;
+    } else if (key->aux.tls_ver >= TLS1_1_VERSION) {
+      iv = AES_BLOCK_SIZE;
+    }
+
+    size_t aes_off = 0;
+    size_t sha_off = SHA_CBLOCK - key->md.num;
+    size_t blocks;
+
+    // TODO: check CPU cap like e_aes_cbc_hmac_sha256.c?
+    if (plen > (sha_off + iv) &&
+        (blocks = (plen - (sha_off + iv)) / SHA_CBLOCK)) {
+      SHA1_Update(&key->md, in + iv, sha_off);
+
+      aesni_cbc_sha1_enc(in, out, blocks, &key->ks,
+                         EVP_CIPHER_CTX_iv_noconst(ctx), &key->md,
+                         in + iv + sha_off);
+      // Update the offset to record and skip the part processed
+      // (encrypted and hashed) by |aesni_cbc_sha1_enc|.
+      blocks *= SHA_CBLOCK;
+      aes_off += blocks;
+      sha_off += blocks;
+      key->md.Nh += blocks >> 29;
+      key->md.Nl += blocks <<= 3;
+      if (key->md.Nl < (unsigned int)blocks) {
+        key->md.Nh++;
+      }
+    } else {
+      sha_off = 0;
+    }
+    sha_off += iv;
+    SHA1_Update(&key->md, in + sha_off, plen - sha_off);
+
+    if (in != out) {
+      OPENSSL_memcpy(out + aes_off, in + aes_off, plen - aes_off);
+    }
+
+    // calculate HMAC and append it to payload.
+    SHA1_Final(out + plen, &key->md);
+    key->md = key->tail;
+    SHA1_Update(&key->md, out + plen, SHA_DIGEST_LENGTH);
+    SHA1_Final(out + plen, &key->md);
+
+    // pad the payload|hmac.
+    plen += SHA_DIGEST_LENGTH;
+    for (unsigned int l = len - plen - 1; plen < len; plen++) {
+      out[plen] = l;
+    }
+    // encrypt HMAC|padding at once.
+    aes_hw_cbc_encrypt(out + aes_off, out + aes_off, len - aes_off, &key->ks,
+                       EVP_CIPHER_CTX_iv_noconst(ctx), 1);
+    return 1;
+  } else {
```

Comment:
I guess the decryption part starts here, but it's difficult to say because both the encryption and decryption are quite long so maybe it's better to move them to separate functions and just call them from this function.